### PR TITLE
*: Support basic aes_encrypt/aes_decrypt function.

### PR DESCRIPTION
**Pull request by shenli**
Support aes_encrypt/aes_decrypt.
We do not support using different aes mode according to  `block_encryption_mode` system variable. We only support aes-128-ecb now.
See https://dev.mysql.com/doc/refman/5.6/en/encryption-functions.html#function_aes-encrypt

@coocood @tiancaiamao @XuHuaiyu PTAL

### CODE REVIEW

File: expression/builtin_encryption.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		// If either function argument is NULL, the function returns NULL.
+		if arg.IsNull() {
+			return
+		}
+	}
+	str := args[0].GetBytes()
+	key := args[1].GetBytes()
+	// TOOD:
+	// MySQL support key with other length, but we still do not know how MySQL handle those case.
+	// We will hanlde those cases latter.
```
Comment (coocood) [inline expression/builtin_encryption.go]: It is implement here
https://github.com/mysql/mysql-server/blob/5.7/mysys_ssl/my_aes.cc

**Comment by shenli** [comment]
@coocood PTAL

**Comment by coocood** [comment]
LGTM

File: expression/builtin.go [hunk 2]
```diff
@@ -284,4 +284,8 @@ var funcs = map[string]functionClass{
 	ast.RowFunc:    &rowFunctionClass{baseFunctionClass{ast.RowFunc, 2, -1}},
 	ast.SetVar:     &setVarFunctionClass{baseFunctionClass{ast.SetVar, 2, 2}},
 	ast.GetVar:     &getVarFunctionClass{baseFunctionClass{ast.GetVar, 1, 1}},
+
+	// encryption and compression functions
+	ast.AesDecrypt: &aesDecryptFunctionClass{baseFunctionClass{ast.AesDecrypt, 2, 3}},
+	ast.AesEncrypt: &aesEncryptFunctionClass{baseFunctionClass{ast.AesDecrypt, 2, 3}},
```
Comment (zimulala) [inline expression/builtin.go]: s/AesDecrypt/AesEncrypt ?

Comment (shenli) [inline expression/builtin.go]: @zimulala PTAL

Comment (hanfei1991) [inline expression/builtin.go]: Decrypt->Encrypt

File: expression/builtin_encryption.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+	key = handleAESKey(key, aes128ecb)
+	crypted, err := encrypt.AESEncryptWithECB(str, key)
+	if err != nil {
+		return d, errors.Trace(err)
+	}
+	d.SetString(string(crypted))
+	return
+}
+
+func handleAESKey(key []byte, mode string) []byte {
```
Comment (hanfei1991) [inline expression/builtin_encryption.go]: add comment for this.

**Comment by shenli** [comment]
@hanfei1991 PTAL

**Comment by zimulala** [comment]
LGTM

File: expression/builtin_encryption.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	// TODO: get key size according to mode
+	keySize := 16
+	rKey := make([]byte, keySize)
+
+	rIdx := 0
+	for _, k := range key {
+		if rIdx == keySize {
```
Comment (hanfei1991) [inline expression/builtin_encryption.go]: Will the rIdx equal to keySize ?

File: expression/builtin_encryption_test.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	c.Assert(err, IsNil)
+	c.Assert(crypt.IsNull(), IsTrue)
+
+	f, err = fc.getFunction(datumsToConstants([]types.Datum{argNull, arg}), s.ctx)
+	crypt, err = f.eval(nil)
+	c.Assert(err, IsNil)
+	c.Assert(crypt.IsNull(), IsTrue)
+}
+
+func (s *testEvaluatorSuite) TestAESDecrypt(c *C) {
```
Comment (hanfei1991) [inline expression/builtin_encryption_test.go]: Maybe you can combine these two tests

File: expression/builtin_encryption.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+		return d, errors.Trace(err)
+	}
+	d.SetString(string(crypted))
+	return
+}
+
+// Transforms an arbitrary long key into a fixed length AES key.
+func handleAESKey(key []byte, mode string) []byte {
+	// TODO: get key size according to mode
+	keySize := 16
```
Comment (zimulala) [inline expression/builtin_encryption.go]: s/16/aes.BlockSize

File: util/encrypt/aes.go [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+func newECBDecrypter(b cipher.Block) cipher.BlockMode {
+	return (*ecbDecrypter)(newECB(b))
+}
+
+// Padding using PKCS5
+// See https://en.wikipedia.org/wiki/Padding_(cryptography)#PKCS7
+func pkcs5Padding(ciphertext []byte, blockSize int) []byte {
+	// The bytes need to padding.
+	padding := blockSize - len(ciphertext)%blockSize
```
Comment (hanfei1991) [inline util/encrypt/aes.go]: If the ciphertext is 16, we still need to padding?

File: util/encrypt/aes.go [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+func AESEncryptWithECB(str, key []byte) ([]byte, error) {
+	cb, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	blockSize := cb.BlockSize()
+	// The str arguments can be any length, and padding is automatically added to
+	// str so it is a multiple of a block as required by block-based algorithms such as AES.
+	// This padding is automatically removed by the AES_DECRYPT() function.
+	// data := padding(str, blockSize)
```
Comment (hanfei1991) [inline util/encrypt/aes.go]: Should we remove this line ?

File: expression/builtin_encryption.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+	rKey := make([]byte, keySize)
+	rIdx := 0
+	for _, k := range key {
```
Comment (tiancaiamao) [inline expression/builtin_encryption.go]: Do you mean?

```
for i:=0; i<len(key); i++ {
    rIdx := i % keySize
    rKey[rIdx] = ^= k
}
```

Where is the reference to this function `handleAESKey` ? Is it mysql's behavior or a standard algorithm?

File: util/encrypt/aes.go [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+		src = src[x.blockSize:]
+		dst = dst[x.blockSize:]
+	}
+}
+
+func newECBDecrypter(b cipher.Block) cipher.BlockMode {
+	return (*ecbDecrypter)(newECB(b))
+}
+
+// Padding using PKCS5
```
Comment (XuHuaiyu) [inline util/encrypt/aes.go]: s/PKCS5/PCKS7 ?

File: util/encrypt/aes.go [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+// See https://en.wikipedia.org/wiki/Padding_(cryptography)#PKCS7
+func pkcs5Padding(ciphertext []byte, blockSize int) []byte {
+	// The bytes need to padding.
+	padding := blockSize - len(ciphertext)%blockSize
+	padtext := bytes.Repeat([]byte{byte(padding)}, padding)
+	return append(ciphertext, padtext...)
+}
+
+// UnPadding using PKCS5
```
Comment (XuHuaiyu) [inline util/encrypt/aes.go]: ditto

File: expression/builtin_encryption.go (hunk 6 above)
Comment (coocood) [inline expression/builtin_encryption.go]: keySize may not be `aes.BlockSize`.

File: expression/builtin_encryption.go (hunk 9 above)
Comment (coocood) [inline expression/builtin_encryption.go]: It's MySQL's behavior.

**Comment by coocood** [comment]
@hanfei1991 @XuHuaiyu PTAL

**Comment by hanfei1991** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
LGTM

**Pull request merged**